exporter/datadogexporter/internal/hostmetadata/provider/provider.go

```diff
@@ -21,16 +21,44 @@ type chainProvider struct {
 	priorityList []string
 }
 
+type providerReply struct {
+	src source.Source
+	err error
+}
+
 func (p *chainProvider) Source(ctx context.Context) (source.Source, error) {
-	for _, source := range p.priorityList {
-		zapProvider := zap.String("provider", source)
+	// Cancel all providers when exiting
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Run all providers in parallel
+	providerReplies := make([]chan providerReply, len(p.priorityList))
+	for i, source := range p.priorityList {
 		provider := p.providers[source]
-		src, err := provider.Source(ctx)
-		if err == nil {
-			p.logger.Info("Resolved source", zapProvider, zap.Any("source", src))
-			return src, nil
+		providerReplies[i] = make(chan providerReply)
+
+		go func(i int, source string) {
```

Comment:
These are not guaranteed to be sending down the channel in the same order that they are created. So they don't really work in terms of "priority". If that's intended, then no worries.

Reply:
Also, I would add a waitgroup and wait for these to finish before returning in this function. Possibly using `defer wg.Wait()`. Just to make sure we're clean here.

Reply:
> These are not guaranteed to be sending down the channel in the same order that they are created. So they don't really work in terms of "priority". If that's intended, then no worries.

The loop below should make it so that even if they don't return in order, the priority will be honored: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a797e5e6e61dc0d1bfda0818e7806bab1b13d013/exporter/datadogexporter/internal/hostmetadata/provider/provider.go#L54-L55

This test explicitly checks this https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a797e5e6e61dc0d1bfda0818e7806bab1b13d013/exporter/datadogexporter/internal/hostmetadata/provider/provider_test.go#L103-L113

p3 will send things down faster than p2, but p2's source is selected.

> Also, I would add a waitgroup and wait for these to finish before returning in this function. Possibly using defer wg.Wait(). Just to make sure we're clean here.

We don't want to wait, we want to return as soon as we have a source. The `defer cancel()` here https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a797e5e6e61dc0d1bfda0818e7806bab1b13d013/exporter/datadogexporter/internal/hostmetadata/provider/provider.go#L31-L32 will make sure that all remaining source providers are cancelled.

Reply:
> We don't want to wait, we want to return as soon as we have a source. The defer cancel() here

Right, but when we do return, we want all goroutines to exit before leaving the function. Otherwise there could be a leak.

Reply:
Regarding the order, that's just plain luck. The quickest provider will send down the channel first. Here's a simple example: https://go.dev/play/p/ATFcpO5a8fL. Feel free to play around with the times.

Reply:
Discussed this offline, waiting may increase running times instead of decreasing, and since this function is called at most once, with 7 providers [https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ac5eb056923[…]e8c92d98/exporter/datadogexporter/internal/hostmetadata/host.go](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ac5eb056923500ca3de6246e3f38d828e8c92d98/exporter/datadogexporter/internal/hostmetadata/host.go#L71), the worst case scenario is that 7 extra goroutines will keep on executing for a bit longer than they should. The `defer cancel()` should make any cancel-honoring provider end when exiting the function.